Managing multiple branches (#15)
inrupt/public-documentation: coding-conventions/git-workflow.md

# Suggested git workflow
The example below shows how you could manage multiple tasks and pull requests in parallel based on the git principles defined in
[Coding standards](coding-standards.md "Coding standards").

| Activity | Git commands | GitHub actions |
| -------- | ------------ | -------------- |
| **Start task 1 on a new branch** | `git checkout main`<br>`git checkout -b branch1` ||
| Commit changes | `git commit -am"Finished work item 1"` | |

> | Check your local branch is up to date | `git checkout main`<br>`git pull`<br>`git checkout branch1`<br>`git rebase main` | |

Reviewer: When your on the branch you want to rebase you can just use: `git fetch` and `git rebase origin/main`

Reviewer 2: What @jholleran describes is what I do when I need to rebase.

Reviewer 3: And me too (not surprisingly, since Jarlath was my Git-Sensei :) !).

| --- | --- | --- |
| **Start task 2 on a new branch** | `git checkout main`<br>`git pull`<br>`git checkout -b branch2` ||
| **Q:** What if you need branch1 code in your work? | Would you merge/rebase that branch in or branch from the branch?|
| Commit changes | `git commit -am"Work in progress on task 2"` ||
| --- | --- | --- |
| **Reviewer suggests changes in PR#1** | `git checkout branch1` ||
| **Q:** What if there are changes on master to incorporate? Do you do another rebase but this is now a public branch so how do you then push? || 
| Commit changes | `git commit -am"Updated task 1"` ||
| Update PR#1 |`git push` ||
| --- | --- | --- |
| **Continue with task 2** | `git checkout branch2` ||
| **Q:** How to you update this to include the branch1 changes? ||
| Commit changes, checking you are up to date with main as before | `git commit -am"Finished task 2"` ||
| Create PR#2 | `git push origin branch2` | GitHub will offer to help you **Compare and pull request**<br>Explain your change and add a reviewer |
| --- | --- | --- |
| **Reviewer approves PR#1** | | Squash and merge into main leaving one commit message |

> | Delete branch1 | `git branch -d branch1`<br>`git push origin --delete branch1` | You can also delete the remote branch in github instead of using the 2nd command |

Reviewer 3: We never manually delete branches (or at least I don't!), we just rely on the "Automatically delete head branches" option under the repo settings (in the "Merge button" section).

```suggestion
| Delete branch1 | `git branch -d branch1`<br>`git push origin --delete branch1` | You can also delete the remote branch in GitHub instead of using the 2nd command |
```

| **Q:** Do we rebase local branch2 on main? ||
| --- | --- | --- |
| **Reviewer suggests changes in PR#2** | `git checkout branch2` ||
| Commit changes | `git commit -am"Updated task 2"` ||
| Update PR#2 |`git push` ||
| --- | --- | --- |
| **Reviewer approves PR#2** | | Squash and merge into main leaving one commit message<br>**Q:** Is there a rebase at this point in GitHub? |
| Delete branch2 | `git branch -d branch2`<br>`git push origin --delete branch2` | You can also delete the remote branch in github instead of using the 2nd command |
| --- | --- | --- |
| And so on... | `git checkout main`<br>`git pull` ||

> ## Other thoughts

Reviewer 3: ```suggestion
## Other questions...
```

> * Should we promote `git switch <branch>` and `git switch -c <branch>` instead of `git checkout <branch>` and `git checkout -b <branch>`

Reviewer 3: ```suggestion
* Should we promote `git switch <branch>` and `git switch -c <branch>` instead of `git checkout <branch>` and `git checkout -b <branch>`?
```

# Suggested git workflow

> The example below shows the steps involved in working on a task and creating a pull requests based on the git principles defined in

Reviewer 3: ```suggestion
The example below shows the steps involved in working on a task and creating a pull request based on the git principles defined in our
```

[Coding standards](coding-standards.md "Coding standards").

Managing multiple PRs in parallel can be tricky. It is strongly recommended to keep tasks independent, each starting from the main branch.

> If a task depends on work done in another branch then the task is blocked until the first PR is merged.

Reviewer 3: I don't think this statement is strictly true (given how powerful and flexible Git is), but it's just that things can get pretty hairy pretty quickly - so I'd just suggest adding a 'generally speaking' here to allude to that:
```suggestion
If a task depends on work done in another branch then, generally speaking, we'd advise considering that task blocked until the other branch is merged.
```

Reviewer 2: I agree 100% with Pat here -- there are things one can do to to handle these cases (I find myself in that circumstance from time to time), but it requires being very careful about rebasing and cherry-picking from other branches. But by describing this as a "general rule", Pat's formulation is great.

| Activity | Git commands | GitHub actions |
| -------- | ------------ | -------------- |
| **Start task 1 on a new branch** | `git checkout main`<br>`git checkout -b branch1` ||
| Commit changes | `git commit -am"Finished work item 1"` | |
| Check your local branch is up to date with main| See [Rebasing a branch](#rebasing-a-branch) | |
| Create pull request PR#1 | `git push origin branch1` | GitHub will offer to help you **Compare and pull request.**<br>Explain your change and add a reviewer. |
| --- | --- | --- |
| **Start task 2 on a new branch** | `git checkout main`<br>`git pull` or `git fetch; git rebase`<br>`git checkout -b branch2` ||
| Commit changes | `git commit -am"Work in progress on task 2"` ||
| --- | --- | --- |

> | **Reviewer suggests changes in PR#1**<br>Avoid rebasing once the PR is in review if you can | `git checkout branch1` ||

Reviewer 3: ```suggestion
| **Reviewer suggests changes in PR#1**<br>(Avoid rebasing once the PR is in review, if you can) | `git checkout branch1` ||
```

| If necessary, update your local branch | See [Rebasing a branch](#rebasing-a-branch) | |
| Commit changes | `git commit -am"Updated task 1"` ||
| Update PR#1 |`git push origin branch1`<br>or, if you rebased, see [Updating a PR](#updating-a-pr)||
| --- | --- | --- |
| **Continue with task 2** | `git checkout branch2` ||
| Commit changes, checking you are up to date with main as before | `git commit -am"Finished task 2"` ||
| Create PR#2 | `git push origin branch2` | GitHub will offer to help you **Compare and pull request.**<br>Explain your change and add a reviewer. |
| --- | --- | --- |

> | **Reviewer approves PR#1** | | Squash and merge into main leaving one commit message. |

Reviewer 3: ```suggestion
| **Reviewer approves PR#1** | | Squash and merge into `main`, leaving one commit message. |
```

| Delete local branch1 | `git branch -d branch1` | If you have set "Automatically delete head branches" GitHub will have deleted the remote branch automatically. |
| --- | --- | --- |
| **Reviewer suggests changes in PR#2** | `git checkout branch2` ||
| If necessary, update your local branch to include the PR#1 commits that are now in main | See [Rebasing a branch](#rebasing-a-branch) | |
| Commit more changes | `git commit -am"Updated task 2"` ||
| Update PR#2 |`git push origin branch2`<br>or, if you rebased, see [Updating a PR](#updating-a-pr)||
| --- | --- | --- |

> | **Reviewer approves PR#2** | | Squash and merge into main leaving one commit message. |

Reviewer 3: ```suggestion
| **Reviewer approves PR#2** | | Squash and merge into `main`, leaving one commit message. |
```

| Delete local branch2 | `git branch -d branch2` | If you have set "Automatically delete head branches" GitHub will have deleted the remote branch automatically. |
| --- | --- | --- |
| And so on... | `git checkout main`<br>`git pull` ||

## Rebasing a branch
You often need to update a branch to include commits that were added to `main` after you created the branch. The steps are along the lines of the following:
```shell
git checkout <branchname>
git fetch
git rebase origin/main
```

This takes the commits in your branch and replays them on top of the newer commits from `main`.
At this point you may have some merge conflicts. Once these have been addressed you need to add them to the commit and continue the rebase.
```shell
git add <conflicted_files>
git rebase --continue
```

## Updating a PR
If you have made changes to a local branch that is tracking a remote branch (PR) there are 2 ways to update it. 
Firstly, if you have not rebased the branch you can update this with:
```shell
git push origin branch1
# or, if upstream is set this can be simplified to
git push
```

> However, if you have rebased, then your commits will have different ids to the commits in the remote branch even 

Reviewer 3: ```suggestion
However, if you have rebased, then your commits will have different ids to the commits in the remote branch, even 
```

> though they look the same. This would make the `push` command above fail. If no-one else is working on that branch and you

Reviewer 3: Just an idea, but it might be nice to insert an example of the `push` command failing, so we can see exactly what it looks like (i.e., it's always reassuring to see the exact error message in the docs that you're facing in your Terminal window!)
```suggestion
though they look the same. This would make the `push` command above fail like this:
<<< Example push failure >>>
If no-one else is working on that branch and you
```
